projects/bms_carrier/test/test_cell_sense.c

```diff
@@ -0,0 +1,265 @@
+#include "cell_sense.h"
+
+#include "bms.h"
+#include "bms_events.h"
+#include "current_sense.h"
+#include "event_queue.h"
+#include "exported_enums.h"
+#include "gpio.h"
+#include "interrupt.h"
+#include "log.h"
+#include "ltc_afe.h"
+#include "ms_test_helpers.h"
+#include "soft_timer.h"
+#include "test_helpers.h"
+
+#define NUM_GOOD_CELL_SENSE_TESTS 3
+#define NUM_DEVICES_TO_TEST 1
+#define NUM_CELLS_TO_TEST 4
+
+#define VOLTAGE_READING_MAX 1337
+#define VOLTAGE_READING_MIN 42
+#define TEMP_READING_MAX 1337
+#define TEMP_READING_MIN 42
+static AfeReadings s_raw_readings = {
+  .voltages = { 42, 69, 420, 1337 },
+  .temps = { 42, 69, 420, 1337 },
+};
+static AfeReadings s_readings;
+
+static bool s_is_charging;
+bool TEST_MOCK(current_sense_is_charging)() {
+  return s_is_charging;
+}
+
+uint8_t s_expected_fault_bitset;
+void TEST_MOCK(fault_bps)(uint8_t fault_bitmask, bool clear) {
+  TEST_ASSERT_EQUAL((s_expected_fault_bitset & fault_bitmask) == 0x0, clear);
+}
+
+static bool s_afe_should_fault;
+bool TEST_MOCK(ltc_afe_process_event)(LtcAfeStorage *afe, const Event *e) {
+  LtcAfeEventList *afe_events = &afe->settings.ltc_events;
+  LtcAfeResultCallback cb;
+  uint16_t *result_arr;
+  if (e->id == BMS_AFE_EVENT_TRIGGER_AUX_CONV) {
+    cb = afe->settings.aux_result_cb;
+    result_arr = s_raw_readings.temps;
+  } else if (e->id == BMS_AFE_EVENT_TRIGGER_CELL_CONV) {
+    cb = afe->settings.cell_result_cb;
+    result_arr = s_raw_readings.voltages;
+  } else {
+    return false;
+  }
+
+  TEST_ASSERT_NOT_NULL(cb);
+  if (s_afe_should_fault)
+    return status_ok(event_raise_priority(EVENT_PRIORITY_HIGHEST, afe_events->fault_event, 0));
+  cb(result_arr, afe->settings.num_cells, afe->settings.result_context);
+  return status_ok(event_raise_priority(EVENT_PRIORITY_HIGHEST, afe_events->callback_run_event, 0));
+}
+
+static LtcAfeStorage s_afe;
+static StatusCode prv_init_ltc(void) {
+  status_ok_or_return(gpio_init());
```

Comment:
why? isn't this guaranteed by other tests?

Reply:
It is guaranteed but I use same pattern as used in the smoketest.
It's not a TEST_ASSERT because it should be guaranteed but I think it's fair to leave it if someone were to run test on h/w.